src/project-converter.ts

```diff
@@ -0,0 +1,159 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Document} from 'polymer-analyzer';
+
+import {ConversionSettings} from './conversion-settings';
+import {DocumentConverter} from './document-converter';
+import {ConversionResult, JsExport} from './js-module';
+import {ConvertedDocumentFilePath, OriginalDocumentUrl} from './urls/types';
+import {UrlHandlerInterface} from './urls/url-handler-interface';
+import {getDocumentUrl} from './urls/util';
+
+/**
+ * ProjectConverter provides the top-level interface for running a project
+ * conversion. convertDocument() should be called to kick off any conversion,
+ * and getResults() should be called once conversion is complete.
+ *
+ * For best results, only one ProjectConverter instance should be needed, so
+ * that it can cache results and avoid duplicate, extraneous document
+ * conversions.
+ *
+ * ProjectConverter is indifferent to the layout of the project, delegating any
+ * special URL handling/resolution to the urlHandler provided to the
+ * constructor.
+ */
+export class ProjectConverter {
+  readonly urlHandler: UrlHandlerInterface;
+  readonly conversionSettings: ConversionSettings;
+
+  /**
+   * A cache of all JS Exports by namespace, to map implicit HTML imports to
+   * explicit named JS imports.
+   */
+  readonly namespacedExports = new Map<string, JsExport>();
+  /**
+   * A cache of all converted documents. Document conversions should be
+   * idempotent, so conversion results can be safely cached.
+   */
+  readonly conversionResults = new Map<OriginalDocumentUrl, ConversionResult>();
+
+  constructor(
+      urlHandler: UrlHandlerInterface, conversionSettings: ConversionSettings) {
+    this.urlHandler = urlHandler;
+    this.conversionSettings = conversionSettings;
+  }
+
+  /**
+   * Convert a document and any of its dependencies. The output format (JS
+   * Module or HTML Document) is determined by whether the file is included in
+   * conversionSettings.includes.
+   */
+  convertDocument(document: Document) {
```

Comment:
Maybe indicate that this `document` variable is always an `HtmlDocument` or `ParsedHtmlDocument` or whatever that class is -- i.e. this is not handed a `JavaScriptDocument` as input right now.

Reply:
Is there an HtmlDocument type? I don't see one exported by the analyzer, and `analysis.getFeatures({kind: 'html-document'})` is typed to return a Set of `Document`s.

Reply:
I just now see that the `Document` model returned in the `getDocument` call of an `Analysis` is basically a wrapper class and that it exposes the typed versions through the `parsedDocument` getter, so you can't express the constraint here through typings AFAICT.  So maybe just a comment, if that.

Reply:
You can definitely check that you're handed an HTML document though, I'd do that.